posteriors/laplace/dense_ggn.py

```diff
@@ -0,0 +1,174 @@
+            as well as auxiliary information.
+        outer_log_likelihood: A function that takes the output of `forward` and batch
+            then returns the log likelihood of the model output,
+            with no auxiliary information.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        GGN Laplace approximation transform instance.
+    """
+    init_fn = partial(init, init_prec=init_prec)
+    update_fn = partial(
+        update, forward=forward, outer_log_likelihood=outer_log_likelihood
+    )
+    return Transform(init_fn, update_fn)
+
+
+@dataclass
+class DenseLaplaceState(TransformState):
+    """State encoding a Normal distribution over parameters,
+    with a dense precision matrix
+
+    Args:
+        params: Mean of the Normal distribution.
+        prec: Precision matrix of the Normal distribution.
+        aux: Auxiliary information from the log_posterior call.
+    """
+
+    params: TensorTree
+    prec: torch.Tensor
+    aux: Any = None
+
+
+def init(
+    params: TensorTree,
+    init_prec: torch.Tensor | float = 0.0,
+) -> DenseLaplaceState:
+    """Initialise Normal distribution over parameters
+    with a dense precision matrix.
+
+    Args:
+        params: Mean of the Normal distribution.
+        init_prec: Initial precision matrix.
+            If it is a float, it is defined as an identity matrix
+            scaled by that float.
+
+    Returns:
+        Initial DenseLaplaceState.
+    """
+
+    if is_scalar(init_prec):
+        num_params = tree_size(params)
+        init_prec = init_prec * torch.eye(num_params, requires_grad=False)
+
+    return DenseLaplaceState(params, init_prec)
+
+
+def update(
+    state: DenseLaplaceState,
+    batch: Any,
+    forward: ForwardFn,
+    outer_log_likelihood: OuterLogProbFn,
+    inplace: bool = False,
+) -> DenseLaplaceState:
+    """Adds GGN matrix over given batch.
+
+    Args:
+        state: Current state.
+        batch: Input data to model.
+        forward: Function that takes parameters and input batch and
+            returns a forward value (e.g. logits), not reduced over the batch,
+            as well as auxiliary information.
+        outer_log_likelihood: A function that takes the output of `forward` and batch
+            then returns the log likelihood of the model output,
+            with no auxiliary information.
+        inplace: If True, then the state is updated in place, otherwise a new state
+            is returned.
+
+    Returns:
+        Updated DenseLaplaceState.
+    """
+    with torch.no_grad(), CatchAuxError():
+        ggn_batch, aux = ggn(
+            partial(forward, batch=batch),
+            partial(outer_log_likelihood, batch=batch),
+            forward_has_aux=True,
+            loss_has_aux=False,
+            normalize=False,
+        )(state.params)
+
+    if inplace:
+        state.prec -= ggn_batch
```

Comment:
why is `- ggn` here and it was `+ laplace`?

Reply:
I would add a comment explaining this in the docstring still. "Because of the way the GGN matrix is defined, note that we update `state.prec` by subtracting the batched ggn to the current `state.prec`."

Reply:
Added a conversion from `outer_log_likelihood` to `outer_loss` which hopefully clarifies things

Reply:
Yep, much better!